library/core/src/slice/mod.rs

```diff
@@ -3169,6 +3187,232 @@ impl<T> [T] {
+    /// ```
+    ///
+    /// Taking the tail of a slice starting at index two (via `2..`):
+    ///
+    /// ```
+    /// #![feature(slice_take)]
+    ///
+    /// let mut slice: &mut [_] = &mut ['a', 'b', 'c', 'd'];
+    /// let mut tail = slice.take_mut(2..).unwrap();
+    ///
+    /// assert_eq!(slice, &mut ['a', 'b']);
+    /// assert_eq!(tail, &mut ['c', 'd']);
+    /// ```
+    ///
+    /// Getting `None` when `range` starts or ends outside of the slice:
+    ///
+    /// ```
+    /// #![feature(slice_take)]
+    ///
+    /// let mut slice: &mut [_] = &mut ['a', 'b', 'c', 'd'];
+    ///
+    /// assert_eq!(None, slice.take_mut(5..));
+    /// assert_eq!(None, slice.take_mut(..5));
+    /// assert_eq!(None, slice.take_mut(..=4));
+    /// let expected: &mut [_] = &mut ['a', 'b', 'c', 'd'];
+    /// assert_eq!(Some(expected), slice.take_mut(..4));
+    /// ```
+    #[inline]
+    #[unstable(feature = "slice_take", issue = "62280")]
+    pub fn take_mut<'a, R: OneSidedRange<usize>>(
+        self: &mut &'a mut Self,
+        range: R,
+    ) -> Option<&'a mut Self> {
+        let (taking_front, split_index) = take_split_point(range)?;
+        if split_index > self.len() {
+            return None;
+        }
+        let original = crate::mem::take(self);
+        let (front, back) = original.split_at_mut(split_index);
+        if taking_front {
+            *self = back;
+            Some(front)
+        } else {
+            *self = front;
+            Some(back)
+        }
+    }
+
+    /// Returns a reference to the first element of the slice,
+    /// and modifies the slice to no longer include this element.
+    ///
+    /// Returns `None` if the slice is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_take)]
+    ///
+    /// let mut slice: &[_] = &['a', 'b', 'c'];
+    /// let first = slice.take_first().unwrap();
+    ///
+    /// assert_eq!(slice, &['b', 'c']);
+    /// assert_eq!(first, &'a');
+    /// ```
+    #[inline]
+    #[unstable(feature = "slice_take", issue = "62280")]
+    pub fn take_first<'a>(self: &mut &'a Self) -> Option<&'a T> {
+        let (first, rem) = self.split_first()?;
+        *self = rem;
+        Some(first)
+    }
+
+    /// Returns a mutable reference to the first element of the slice,
+    /// and modifies the slice to no longer include this element.
+    ///
+    /// Returns `None` if the slice is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_take)]
+    ///
+    /// let mut slice: &mut [_] = &mut ['a', 'b', 'c'];
+    /// let first = slice.take_first_mut().unwrap();
+    /// *first = 'd';
+    ///
+    /// assert_eq!(slice, &['b', 'c']);
+    /// assert_eq!(first, &'d');
+    /// ```
+    #[inline]
+    #[unstable(feature = "slice_take", issue = "62280")]
+    pub fn take_first_mut<'a>(self: &mut &'a mut Self) -> Option<&'a mut T> {
+        let (first, rem) = mem::take(self).split_first_mut()?;
```

Comment:
Is `mem::take` call here really necessary?

Reply:
`split_first_mut` takes ownership of the `&mut [T]` it's called on, and in this context we only have a `&mut &mut [T]`.

Reply:
Hmm, what do you mean by "it takes ownership of the `&mut [T]`"? An `&mut [T]` is a borrow TMK.

Reply:
The double `&mut` makes this a bit confusing. Perhaps a better way of phrasing it is that the `split_mut` methods return unique references to the contents of the input slice, and therefore they can't coexist with `self` as long as `self` is pointing to that data as well. We need `self` to point to something else first, which is what `mem::take` does for us here.

Reply:
Your code is equivalent to:

```rust
use std::mem;

pub fn take_first_mut<'a, T>(slice: &mut &'a mut [T]) -> Option<&'a mut T> {
    let (first, rem) = mem::take(slice).split_first_mut()?;
    *slice = rem;
    Some(first)
}
```

Just adding a `'a` to the outer reference makes it compile without `mem::take`:

```rust
pub fn take_first_mut<'a, T>(slice: &'a mut &'a mut [T]) -> Option<&'a mut T> {
    let (first, rem) = slice.split_first_mut()?;
    *slice = rem;
    Some(first)
}
```

IIUC in this case `&mut &'a mut Self` is equivalent to `&'a mut &'a mut Self` because of how lifetime elision works. So (again, IIUC) adding the `'a` to the outer reference should not change the API of the function at all.

Reply:
>IIUC in this case `&mut &'a mut Self` is equivalent to `&'a mut &'a mut Self` because of how lifetime elision works.

If the two were equivalent, changing it from one to the other wouldn't have made non-compiling code compile 🙂

It's essential here that the return value of these methods borrows from the "inner" slice, which could have a very long lifetime, and not from the "outer" reference to it, which could be much more short-lived. Changing the type of `slice` to `&'a mut &'a mut [T]` would needlessly restrict the lifetime of the return value to the outer lifetime of the input value.

Here's some code that doesn't compile with the second version of `take_first_mut` you posted ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=066615fe88bd1e56c1f2ff7a97da73a2)):

```rust
let mut a: &mut [u8] = &mut [1, 2, 3];
println!("{:?}", take_first_mut(&mut a));
println!("{:?}", &mut a);
```

Reply:
Hmm, okay. Then what is the lifetime of the outer reference?